x-pack/plugins/apm/public/components/shared/environment_select/index.tsx

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { i18n } from '@kbn/i18n';
+import React, { useState, useCallback } from 'react';
+import { debounce } from 'lodash';
+import { EuiComboBox, EuiComboBoxOptionOption } from '@elastic/eui';
+import { getEnvironmentLabel } from '../../../../common/environment_filter_values';
+import { SERVICE_ENVIRONMENT } from '../../../../common/elasticsearch_fieldnames';
+import { FETCH_STATUS, useFetcher } from '../../../hooks/use_fetcher';
+import { Environment } from '../../../../common/environment_rt';
+
+export function EnvironmentSelect({
+  environment,
+  environmentOptions,
+  status,
+  serviceName,
+  start,
+  end,
+  onChange,
+}: {
+  environment: Environment;
+  environmentOptions: Array<EuiComboBoxOptionOption<string>>;
+  status: FETCH_STATUS;
+  serviceName?: string;
+  start?: string;
+  end?: string;
+  onChange: (value: string) => void;
+}) {
+  const [searchValue, setSearchValue] = useState('');
+  const defaultOption: EuiComboBoxOptionOption<string> = {
+    value: environment,
+    label: getEnvironmentLabel(environment),
+  };
+
+  const [selectedOptions, setSelectedOptions] = useState(
+    defaultOption ? [defaultOption] : []
+  );
+
+  const handleChange = useCallback(
+    (changedOptions: Array<EuiComboBoxOptionOption<string>>) => {
+      setSelectedOptions(changedOptions);
+
+      if (changedOptions.length === 1 && changedOptions[0].value) {
+        onChange(changedOptions[0].value);
+      }
+    },
+    [onChange]
+  );
```

Comment:
I'm not sure if we need state here? I would expect this component to take an `environment` prop and an `onChange` prop, and have the consuming component handle state changes / synchronisation with the URL.

In fact, I think there's a bug: if you select an environment, and then use browser navigation to go back, it doesn't update the selected options.

Reply:
good catch 😮  I just tried it and there is a bug as you described 🐛

Reply:
fixed here: https://github.com/elastic/kibana/pull/126679/commits/31e00364e3f99960b50a3a157b1c8d2ebedb83ac